feat: Provide support for batch scraping

hkfgo opened the pull request:
Hey Tom, hope all is well!  

I've attached this diagram to illustrate my proposed change. 
<img width="1790" alt="Screenshot 2024-03-22 at 10 23 21 PM" src="https://github.com/tomkerkhove/promitor/assets/1821158/3af9d496-3a4f-41ef-a827-21c0e0bcbf46">

### How to Batch? 
Scrape jobs are currently defined per <resource, metric>. I'm taking the simplest approach to just batch resources looking for the same metric. 

Max batch size will be a configurable parameter, under runtime configuration. 
### Batch Job Execution Flow
![Screenshot 2024-08-21 at 5 37 25 PM](https://github.com/user-attachments/assets/e809782b-9e15-4c47-b757-997b3aaafd4d)

### Changes to Internal Metrics 
Internal Promitor metrics will need to be modified under batch scraping mode. 
`promitor_scrape_success` and `promitor_scrape_error` metrics should be appended with an additional tag `batched`, to indicate the mode of operation(0/1) 
A histogram metric `promitor_batch_size` should be added to track actual batch sizes 
`promitor_ratelimit_arm_throttled`/`promitor_ratelimit_arm` metrics will no longer be published, since batch scraping does not go through ARM.  

### Other considerations
Batch scraping will not be free anymore but comes with other benefits(like no longer being subject to ARM throttling). This should be documented for all users 

Implements #2284

hkfgo commented on src/Promitor.Core.Scraping/Configuration/Model/Metrics/ScrapeDefinitionBatch.cs:
```diff
@@ -0,0 +1,65 @@
+
+    public TimeSpan AggregationInterval{ get; }
+
+    public override int GetHashCode()
```
Do you think this is a sane way to implement HashCode?

tomkerkhove replied:
Do we need one? If not, let's remove it

hkfgo replied:
The idea I have is to use hash code computation to enforce batch restrictions. This part should be sane. 

How I'm choosing to implement hash code is hashing a unique string representation of `ScrapeDefinitionBatch`. Curious what you think/can think of something better?

tomkerkhove replied:
If we really need one, then this is OK for me

tomkerkhove commented:
Thanks! Is it safe to say that the top of the diagram is current and bottom shows these changes?

tomkerkhove commented on src/Promitor.Core.Scraping/Configuration/Model/Metrics/BatchScrapeDefinition.cs:
```diff
@@ -0,0 +1,44 @@
+        /// </param>
+        public BatchScrapeDefinition(List<ScrapeDefinition<TResourceDefinition>> groupedScrapeDefinitions, ScrapeDefinitionBatchProperties scrapeDefinitionBatchProperties)
+        {
+            Guard.NotNull(groupedScrapeDefinitions, nameof(groupedScrapeDefinitions));
+            Guard.NotNull(groupedScrapeDefinitions, nameof(scrapeDefinitionBatchProperties));
```
Let's guard for things that have no entries

tomkerkhove replied:
```suggestion
            Guard.NotNull(groupedScrapeDefinitions, nameof(groupedScrapeDefinitions));
            Guard.NotNull(scrapeDefinitionBatchProperties, nameof(scrapeDefinitionBatchProperties));
```

tomkerkhove commented on src/Promitor.Core.Scraping/Configuration/Model/Metrics/ScrapeDefinition.cs:
```diff
@@ -94,5 +94,14 @@ public class ScrapeDefinition<TResourceDefinition> where TResourceDefinition : c
             return AzureMetricConfiguration?.Aggregation?.Interval;
         }
+
+        public ScrapeDefinitionBatchProperties buildPropertiesForBatch() { 
```
Massive nit
```suggestion
        public ScrapeDefinitionBatchProperties BuildScrapingBatchInfo() { 
```

tomkerkhove commented on src/Promitor.Core.Scraping/Configuration/Model/Metrics/ScrapeDefinitionBatch.cs:
```diff
@@ -0,0 +1,65 @@
+    /// <summary>
+    /// Equality comparison override in case of hash collision
+    /// </summary>
+    public override bool Equals(object obj)
```
Same here, let's remove if we don't need it

tomkerkhove commented on src/Promitor.Core.Scraping/Scraper.cs:
```diff
@@ -103,10 +103,60 @@ public async Task ScrapeAsync(ScrapeDefinition<IAzureResourceDefinition> scrapeD
+                    aggregationType,
+                    aggregationInterval.Value);
+                
+                foreach (int i in Enumerable.Range(0, scrapedMetricResult.Count))
+                {
+                    var scrapedMetricResult = scrapedMetricResult[i];
+                    var scrapeDefinition = scrapeDefinitions[i];
+                    LogMeasuredMetrics(scrapeDefinition, scrapedMetricResult, aggregationInterval);
+
+                    await _metricSinkWriter.ReportMetricAsync(scrapeDefinition.PrometheusMetricDefinition.Name, scrapeDefinition.PrometheusMetricDefinition.Description, scrapedMetricResult);
+
+                    await ReportScrapingOutcomeAsync(scrapeDefinition, isSuccessful: true, isBatchJob: true) ;
+                }
```
We may need to fan this out to running tasks in parallel for every result, but can do that in a later update

hkfgo replied:
Yes agreed, there are potentially a lot of results to stuff into a single thread. While we are on this topic, I believe timing metric(s) for task execution would be great to add. You know, so we can measure outcomes quantitively

tomkerkhove reviewed (commented):
Good start, added a few remarks but no concerns!

hkfgo commented:
Hey I'm revisiting this PR now since SDK migration is almost across the finish line. Do you know if the batch scraping API is available in sovereign clouds and what URLs we should use? We are particularly interested in the US Gov cloud. I can't find any information online(Azure doesn't really mentioned the existence of sovereign clouds to begin with)

tomkerkhove commented:
> Do you know if the batch scraping API is available in sovereign clouds and what URLs we should use?

I do not know, sorry! What is the URL for public cloud?

tomkerkhove commented:
- US Government cloud : <region>.metrics.monitor.azure.us
  - Token audience: https://metrics.monitor.azure.us
- China cloud : <region>.metrics.monitor.azure.cn
  - Token audience: https://metrics.monitor.azure.cn

hkfgo commented:
> * https://metrics.monitor.azure.cn

Awesome! How did you find this out?

tomkerkhove commented:
@hkfgo Do you know if you'll have time to finalize this PR next week? I'd like to cut a release next week.

hkfgo commented:
> @hkfgo Do you know if you'll have time to finalize this PR next week? I'd like to cut a release next week.

We should plan for batch scraping to be on the release after this coming one. There are still a decent amount of work left and I'm not able to work on this full time. 

I do have time to get SDK migration done though. In that case we should have something major-ish to release :)

tomkerkhove commented:
That's OK - Thank you!

tomkerkhove commented:
@hkfgo Any thoughts on this PR? Are you able to continue working on it?

hkfgo commented:
> @hkfgo Any thoughts on this PR? Are you able to continue working on it?

Hey I got caught up in some other work and took a few week PTO. I'm back working on it now :)

hkfgo commented:
Any thoughts on this btw: https://github.com/tomkerkhove/promitor/pull/2459#discussion_r1619981136?

hkfgo commented:
Hey Tom, since you have a direct line to MSFT people, I have an issue w.r.t response data modeling in `Azure.Monitor.Query` you could potentially help with :)  

As you may know, response for a batch query is, at the highest level, separated into a list where each item contains metric results associated with a unique resource. For each item,`resourceid` is included as a JSON attribute when the request is made through plain HTTP REST/ E.g. 
`"resourceid": "/subscriptions/be4fa4c1-8513-4755-95f9-a8627814ca36/resourceGroups/us2-main-ca-rto-messaging/providers/Microsoft.Cache/Redis/axon-messaging-us2"` 

The issue I have now is this attribute is missing in the most recent `Azure.Monitor.Query` release. See their [data model class](https://github.com/Azure/azure-sdk-for-net/blob/05afb36edea3832eb86c3d169149bacd33b18436/sdk/monitor/Azure.Monitor.Query/src/Models/MetricResult.cs#L10). 

This is not a blocker I believe, since resource id can still be parsed out of the `id` attribute of the `MetricResult`. e.g. 
first of the id attribute is the resource ID`"id": "/subscriptions/be4fa4c1-8513-4755-95f9-a8627814ca36/resourceGroups/us2-main-ca-rto-messaging/providers/Microsoft.Cache/Redis/axon-messaging-us2/providers/Microsoft.Insights/metrics/errors"`But such string processing is not ideal of course  

I'll open a GitHub issue too once I get around to it

hkfgo commented:
I got some good news to share :) I observed. in the CI pipeline, the batch scrape jobs working end-to-end. That included grouping to batches, job dispatch, and results processing. While there are still a few thing to work out, I think what I have now is ready for another look! Would you mind doing another round of review? 

Since this is almost a XL-sized PR, I've attached another diagram to help you(and myself) to understand execution flow of these batch jobs.

hkfgo commented:
Also, these are some work items that I've excluded from this PR, in order to keep its size manageable. I believe it makes more sense to do these as follow-ups. Let me know if you agree?
- [ ]  Deal with potential JSON data volume issues with multi-threaded result processing. Per your comment here: https://github.com/tomkerkhove/promitor/pull/2459#discussion_r1551478172
- [ ]  Support for multi-region batch scraping. Since the batch endpoint is now regional(as opposed to global to cloud), we need to do additional work on Promitor's end to support use cases where resources are spread across regions.  
- [ ]  Support batching for log analytics - batch scraping also exists for LogAnalytics, but I want to keep this PR to metrics only
- [ ]  Modify CI to test both single-resource and batch scraping - I think it makes sense to add another runtime config for batch scraping. This way we can test both single-resource and batch scraping in the CI pipeline.

tomkerkhove commented:
Sorry for the delay, I have been out sick & traveling. Will get back to this asap

tomkerkhove commented:
> Also, these are some work items that I've excluded from this PR, in order to keep its size manageable. I believe it makes more sense to do these as follow-ups. Let me know if you agree?
> 
> * [ ]   Deal with potential JSON data volume issues with multi-threaded result processing. Per your comment here: [feat: Provide support for batch scraping #2459 (comment)](https://github.com/tomkerkhove/promitor/pull/2459#discussion_r1551478172)
> * [ ]   Support for multi-region batch scraping. Since the batch endpoint is now regional(as opposed to global to cloud), we need to do additional work on Promitor's end to support use cases where resources are spread across regions.
> * [ ]   Support batching for log analytics - batch scraping also exists for LogAnalytics, but I want to keep this PR to metrics only
> * [ ]   Modify CI to test both single-resource and batch scraping - I think it makes sense to add another runtime config for batch scraping. This way we can test both single-resource and batch scraping in the CI pipeline.

Agreed!

tomkerkhove commented:
> I got some good news to share :) I observed, in the CI pipeline, the batch scrape jobs working end-to-end. That included grouping to batches, job dispatch, and results processing. While there are still a few thing to work out, I think what I have now is ready for another look! Would you mind doing another round of review?
> 
> Since this is almost a XL-sized PR, I've attached another diagram to help you(and myself) to understand execution flow of these batch jobs.

Is the goal to have it merged or just WIP? I'm asking because you mention some things are done in a seperate PR so want to et the expectations right.

If you want to merge it, please publish the PR

hkfgo commented:
> > I got some good news to share :) I observed, in the CI pipeline, the batch scrape jobs working end-to-end. That included grouping to batches, job dispatch, and results processing. While there are still a few thing to work out, I think what I have now is ready for another look! Would you mind doing another round of review?
> > Since this is almost a XL-sized PR, I've attached another diagram to help you(and myself) to understand execution flow of these batch jobs.
> 
> Is the goal to have it merged or just WIP? I'm asking because you mention some things are done in a seperate PR so want to et the expectations right.
> 
> If you want to merge it, please publish the PR

Right to be clear the goal is to merge this PR, and then do all the things I mentioned as follow-up. I'll create separate issues to track them. I've taken the "draft" label off

tomkerkhove commented:
Thanks, I'll check tomorrow!

hkfgo commented on src/Promitor.Integrations.AzureMonitor/HttpPipelinePolicies/LogOutgoingRequestsPolicy.cs:
```diff
@@ -0,0 +1,54 @@
+using Microsoft.Extensions.Logging;
+
+namespace Promitor.Integrations.AzureMonitor.HttpPipelinePolicies{
+    public class ModifyOutgoingAzureMonitorRequestsPolicy : HttpPipelinePolicy
```
This is an insane hack needed because the SDK doesn't work like it's supposed to when you query over a time range :/ https://github.com/Azure/azure-sdk-for-net/issues/40047

tomkerkhove replied:
Can we add a comment to this bug and why we have it?

Thanks for working around it btw!

tomkerkhove commented on src/Promitor.Agents.Scraper/Scheduling/ResourcesScrapingJob.cs:
```diff
@@ -251,22 +251,63 @@ private void GetResourceScrapeDefinition(IAzureResourceDefinition resourceDefini
+        {   
             var tasks = new List<Task>();
+            var batchScrapingEnabled = this._azureMonitorIntegrationConfiguration.Value.MetricsBatching?.Enabled ?? false;
+            Logger.LogInformation("Parsed batch config: {Enabled}, {BatchSize}",  this._azureMonitorIntegrationConfiguration.Value.MetricsBatching.Enabled, this._azureMonitorIntegrationConfiguration.Value.MetricsBatching.MaxBatchSize);
```
This can cause null reference

hkfgo replied:
Moved this down to only log when batching is enabled

tomkerkhove commented on src/Promitor.Agents.Scraper/Scheduling/ResourcesScrapingJob.cs:
```diff
@@ -251,22 +251,63 @@ private void GetResourceScrapeDefinition(IAzureResourceDefinition resourceDefini
             var tasks = new List<Task>();
+            var batchScrapingEnabled = this._azureMonitorIntegrationConfiguration.Value.MetricsBatching?.Enabled ?? false;
+            Logger.LogInformation("Parsed batch config: {Enabled}, {BatchSize}",  this._azureMonitorIntegrationConfiguration.Value.MetricsBatching.Enabled, this._azureMonitorIntegrationConfiguration.Value.MetricsBatching.MaxBatchSize);
+            Logger.LogInformation("Parsed SDK runtime config {Enabled}",  this._azureMonitorIntegrationConfiguration.Value.UseAzureMonitorSdk);
```
Can we consolidate this with the log above?

hkfgo replied:
Don't think we need this actually, it was just a logging statement I used for debugging

tomkerkhove commented on src/Promitor.Core.Scraping/AzureMonitorScraper.cs:
```diff
@@ -73,6 +82,72 @@ protected override async Task<ScrapeResult> ScrapeResourceAsync(string subscript
+                        resourceName: scrapeDefinition.Resource.ResourceName
+                    ); // the resource definition attached is missing some attributes, filling them in here
+                    Logger.LogWarning("Caching resource group {Group} for {ResourceId}",  resourceDefinitionToCache.ResourceGroupName, resourceUri);
+                    _resourceDefinitions.TryAdd(resourceUri, resourceDefinitionToCache);
```
Instead of contains key, we can consider using addorupdate instead as well though

tomkerkhove commented on src/Promitor.Core.Scraping/AzureMonitorScraper.cs:
```diff
@@ -73,6 +82,72 @@ protected override async Task<ScrapeResult> ScrapeResourceAsync(string subscript
+                }
+            }
+
+            var metricFilter = DetermineMetricFilter(metricName, (TResourceDefinition) batchScrapeDefinition.ScrapeDefinitions[0].Resource);
```
Mind elaborating on why we are taking the first entry?

hkfgo replied:
It's a bit of copy-pasta actually. On second thought, it doesn't make sense to construct resource-specific filters when batch scraping.

tomkerkhove commented on src/Promitor.Core.Scraping/AzureMonitorScraper.cs:
```diff
@@ -73,6 +82,72 @@ protected override async Task<ScrapeResult> ScrapeResourceAsync(string subscript
+
+            var metricFilter = DetermineMetricFilter(metricName, (TResourceDefinition) batchScrapeDefinition.ScrapeDefinitions[0].Resource);
+            var metricLimit = batchScrapeDefinition.ScrapeDefinitionBatchProperties.AzureMetricConfiguration.Limit;
+            var dimensionNames = DetermineMetricDimensions(metricName, (TResourceDefinition) batchScrapeDefinition.ScrapeDefinitions[0].Resource, batchScrapeDefinition.ScrapeDefinitionBatchProperties.AzureMetricConfiguration); // TODO: resource definition doesn't seem to be used, can we remove it from function signature?  
```
Is this TODO still applicable?

hkfgo replied:
Somewhat.. It's more to write down an observation I made

tomkerkhove commented on src/Promitor.Core.Scraping/Batching/AzureResourceDefinitionBatching.cs:
```diff
@@ -0,0 +1,39 @@
+        /// </summary>
+        private static List<List<ScrapeDefinition<IAzureResourceDefinition>>> SplitScrapeDefinitionBatch(List<ScrapeDefinition<IAzureResourceDefinition>> batchToSplit, int maxBatchSize, CancellationToken cancellationToken) 
+        {
+            int numNewGroups = (batchToSplit.Count - 1) / 50 + 1;
```
Mind elaborating these magical numbers?

hkfgo replied:
I hardcoded 50(the Azure Monitor limit) instead of using `maxBatchSize`. Fixed now :)

tomkerkhove commented on src/Promitor.Core/Extensions/MeasureMetricExtensions.cs:
```diff
@@ -0,0 +1,19 @@
+        /// </summary>
+        /// <param name="azureCloud">Microsoft Azure cloud</param>
+        /// <returns>Azure environment information for specified cloud</returns>
+        public static ResourceAssociatedMeasuredMetric WithResourceIdAssociation(this MeasuredMetric measuredMetric, string resourceId)
```
Let's make sure we have unit tests for this

tomkerkhove commented on src/Promitor.Integrations.AzureMonitor/HttpPipelinePolicies/LogOutgoingRequestsPolicy.cs:
```diff
@@ -0,0 +1,54 @@
+            var query = System.Web.HttpUtility.ParseQueryString(uriBuilder.Query);
+            foreach (var param in paramNames)
+            {
+                _logger.LogWarning("Original URI param {param} is {value}", param, query[param]);
```
Will this be fore every request? If so, maybe use info given we're handling it?

hkfgo replied:
I changed the pipeline policy to outgoing request params, so this is no longer relevant. This policy will be used to modify batch requests.

tomkerkhove commented on src/Promitor.Integrations.Sinks.OpenTelemetry/Collectors/OpenTelemetrySystemMetricsSink.cs:
```diff
@@ -20,5 +20,10 @@ public async Task WriteGaugeMeasurementAsync(string name, string description, do
+
+        public Task WriteHistogramMeasurementAsync(string name, string description, double value, Dictionary<string, string> labels, bool includeTimestamp)
+        {
+            throw new System.NotImplementedException("Histogram measurement not yet supported in OpenTelemetry sink");
```
Let's not throw but log or ignore instead; this would break end-users using this sink

tomkerkhove commented on src/Promitor.Integrations.Sinks.Prometheus/Collectors/AzureScrapingSystemMetricsPublisher.cs:
```diff
@@ -48,5 +48,14 @@ public async Task WriteGaugeMeasurementAsync(string name, string description, do
+        
+        public async Task WriteHistogramMeasurementAsync(string name, string description, double value, Dictionary<string, string> labels)
+        {
+            throw new System.NotImplementedException();
```
Let's not throw but log or ignore instead; this would break end-users using this sink

tomkerkhove commented on src/Promitor.Integrations.Sinks.Prometheus/Collectors/AzureScrapingSystemMetricsPublisher.cs:
```diff
@@ -48,5 +48,14 @@ public async Task WriteGaugeMeasurementAsync(string name, string description, do
+        }
+        public async Task WriteHistogramMeasurementAsync(string name, string description, double value, Dictionary<string, string> labels, bool includeTimestamp)
+        {
+            throw new System.NotImplementedException();
```
Let's not throw but log or ignore instead; this would break end-users using this sink

tomkerkhove commented on src/Promitor.Integrations.Sinks.Prometheus/Collectors/AzureScrapingSystemMetricsPublisher.cs:
```diff
@@ -48,5 +48,14 @@ public async Task WriteGaugeMeasurementAsync(string name, string description, do
+        {
+            throw new System.NotImplementedException();
+        }
+        public async Task WriteHistogramMeasurementAsync(string name, string description, double value, Dictionary<string, string> labels, bool includeTimestamp)
```
Don't need async here

tomkerkhove commented on src/Promitor.Integrations.Sinks.Prometheus/Collectors/AzureScrapingSystemMetricsPublisher.cs:
```diff
@@ -48,5 +48,14 @@ public async Task WriteGaugeMeasurementAsync(string name, string description, do
             await _systemMetricsPublisher.WriteGaugeMeasurementAsync(name, description, value, labels, includeTimestamp);
         }
+        
+        public async Task WriteHistogramMeasurementAsync(string name, string description, double value, Dictionary<string, string> labels)
```
Don't need async here

hkfgo replied:
Hm are you sure? I thought `async` is needed when you await in the body?

tomkerkhove commented on src/Promitor.Integrations.Sinks.Prometheus/Collectors/PrometheusSystemMetricsSink.cs:
```diff
@@ -36,5 +36,23 @@ public Task WriteGaugeMeasurementAsync(string name, string description, double v
+        /// <param name="includeTimestamp">Indication whether or not a timestamp should be reported</param>
+        public Task WriteHistogramMeasurementAsync(string name, string description, double value, Dictionary<string, string> labels, bool includeTimestamp)
+        {
+            var orderedLabels = labels.OrderByDescending(kvp => kvp.Key).ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
```
What if labels are null?

hkfgo replied:
Added a null check for this histogram and the gauge function above

tomkerkhove commented on src/Promitor.Integrations.Sinks.Prometheus/Collectors/PrometheusSystemMetricsSink.cs:
```diff
@@ -36,5 +36,23 @@ public Task WriteGaugeMeasurementAsync(string name, string description, double v
+        {
+            var orderedLabels = labels.OrderByDescending(kvp => kvp.Key).ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+
+            // TODO: are histogram instruments created on every invocation? Would that interfere with correctness? 
```
Is this still relevant?

hkfgo replied:
Well, I was worried if this would create a new instrument and reset histogram counts every time, but it didn't seem to be the case. I saw histogram values accumulate properly when I ran this in our staging environment

hkfgo replied:
So no it's no longer relevant. I removed it

tomkerkhove reviewed (commented):
Added a few comments but generally looks good and trust you have tested this properly!

Would you mind opening a PR for our docs on https://github.com/promitor/docs as well please?

hkfgo commented:
> Added a few comments but generally looks good and trust you have tested this properly!
> 
> Would you mind opening a PR for our docs on https://github.com/promitor/docs as well please?



> Added a few comments but generally looks good and trust you have tested this properly!
> 
> Would you mind opening a PR for our docs on https://github.com/promitor/docs as well please?



> Added a few comments but generally looks good and trust you have tested this properly!
> 
> Would you mind opening a PR for our docs on https://github.com/promitor/docs as well please?

I have indeed done a fair amount of testing. In CI, I configured the scraper to run under batch scraping mode and it looked to have scraped identical metrics as single-resource scraping through ARM. I've also tested this in our staging environment, with Azure Redis and AzureSQL deployments. The transition going from single-resource scraping to batched scraping was fairly seamless, and the new histogram metric to track batch size appears to work as well. 

I can use some help getting the CI pipeline unblocked btw. As you may have noticed, it's broken for both scraper and resource discovery agent. The resource discovery tests appear to be discovering more resources than expected:
```
[xUnit.net 00:00:06.39]     Promitor.Tests.Integration.Services.ResourceDiscovery.ResourceDiscoveryTests.ResourceDiscoveryV1_GetWithFilterOnTwoRegions_ReturnsExpectedAmount [FAIL]
  Failed Promitor.Tests.Integration.Services.ResourceDiscovery.ResourceDiscoveryTests.ResourceDiscoveryV2_GetWithFilterOnMultipleValueInstanceTag_ReturnsExpectedAmount [13 ms]
  Error Message:
   Assert.Equal() Failure: Values differ
Expected: 405
Actual:   409
```
Can you quickly check if there's a real change in number of logic app instances? If so we may need to update those tests 

Scraper agent is failing too unfortunately :/ The timing(plus the fact that only OpenTelemetry Collector tests are failing) coincided with a [new release]((https://github.com/open-telemetry/opentelemetry-collector-releases/releases/tag/v0.104.0). Those tests worked when I reverted to an older version(0.103.0). I'm playing around with different configs to make tests work with the latest versions

hkfgo commented:
One more thing - if we merge this PR, I'd have to run the CI in single-scraping mode. This is because some resources are in different Azure regions and batch scraping only works single-region right now. Multi-region support is a follow-up I'll take on later(will create an issue for it). 

If that sounds good to you, let me clean up the code and address your comments. Looking forward to getting this merged soon :)

tomkerkhove commented:
Sorry for the delay, checking now!

tomkerkhove commented on .github/workflows/templates-build-push-image.yml:
```diff
@@ -1,5 +1,6 @@
+name: Build and Push (Linux)
```
Can you please revert this?

tomkerkhove commented on config/opentelemetry-collector/collector-config.yaml:
```diff
@@ -2,7 +2,8 @@ receivers:
     protocols:
       grpc:
-
+    tls:
```
Can we revert this?

hkfgo replied:
Yes. I was (unsuccessfully) trying to fix the OTEL Collector CI :/

tomkerkhove commented on src/Promitor.Core.Scraping/LogAnalyticsScraper.cs:
```diff
@@ -49,5 +52,14 @@ private Dictionary<string, string> DetermineMetricLabels(LogAnalyticsResourceDef
+
+        protected override async Task<List<ScrapeResult>> BatchScrapeResourceAsync(string subscriptionId, BatchScrapeDefinition<IAzureResourceDefinition> batchScrapeDefinition, PromitorMetricAggregationType aggregationType, TimeSpan aggregationInterval)
+        {
+            // TODO: these just dispatch and await on tasks that do single-resource scraping. Implement integration with Log Analytics batch endpoint  
```
Think this is redundant now

tomkerkhove commented on src/Promitor.Integrations.AzureMonitor/Extensions/MetricResultExtension.cs:
```diff
@@ -0,0 +1,43 @@
+        static Regex resourceIdRegex = new Regex(resourceIdPattern, RegexOptions.Compiled);
+        
+        public static string ParseResourceIdFromResultId(this MetricResult metricResult) 
+        {
```
Is this still required?

hkfgo replied:
Yes unfortunately, because the SDK does not have a `ResourceID` field that's part of the JSON response. I've added comment in this class to explain the situation

tomkerkhove commented:
> > Added a few comments but generally looks good and trust you have tested this properly!
> > Would you mind opening a PR for our docs on [promitor/docs](https://github.com/promitor/docs) as well please?
> 
> I have indeed done a fair amount of testing. In CI, I configured the scraper to run under batch scraping mode and it looked to have scraped identical metrics as single-resource scraping through ARM. I've also tested this in our staging environment, with Azure Redis and AzureSQL deployments. The transition going from single-resource scraping to batched scraping was fairly seamless, and the new histogram metric to track batch size appears to work as well.
> 
> I can use some help getting the CI pipeline unblocked btw. As you may have noticed, it's broken for both scraper and resource discovery agent. The resource discovery tests appear to be discovering more resources than expected:
> 
> ```
> [xUnit.net 00:00:06.39]     Promitor.Tests.Integration.Services.ResourceDiscovery.ResourceDiscoveryTests.ResourceDiscoveryV1_GetWithFilterOnTwoRegions_ReturnsExpectedAmount [FAIL]
>   Failed Promitor.Tests.Integration.Services.ResourceDiscovery.ResourceDiscoveryTests.ResourceDiscoveryV2_GetWithFilterOnMultipleValueInstanceTag_ReturnsExpectedAmount [13 ms]
>   Error Message:
>    Assert.Equal() Failure: Values differ
> Expected: 405
> Actual:   409
> ```
> 
> Can you quickly check if there's a real change in number of logic app instances? If so we may need to update those tests
> 
> Scraper agent is failing too unfortunately :/ The timing(plus the fact that only OpenTelemetry Collector tests are failing) coincided with a [new release](([open-telemetry/opentelemetry-collector-releases@`v0.104.0` (release)](https://github.com/open-telemetry/opentelemetry-collector-releases/releases/tag/v0.104.0)). Those tests worked when I reverted to an older version(0.103.0). I'm playing around with different configs to make tests work with the latest versions

Yeah, valid point - I have opened https://github.com/tomkerkhove/promitor/pull/2552

hkfgo commented on src/Promitor.Core.Scraping/Configuration/Model/AzureMetricConfiguration.cs:
```diff
@@ -45,5 +45,20 @@ public class AzureMetricConfiguration
         }
+
+         // A unique string to represent this Azure metric and its configured dimensions
+        public string ToUniqueStringRepresentation()
```
Unit tests added for batching heavior

hkfgo commented:
Alright, I've cleaned up the code and addressed all comments! This PR should be in a mergeable state now, besides that the Scraper CI is still failing. I created an [issue](https://github.com/tomkerkhove/promitor/pull/2554) to track that. I think the tests are just flaky since you can see them passing on recent runs: https://dev.azure.com/tomkerkhove/Promitor/_build?definitionId=50&_a=summary

hkfgo commented:
Oh and I'm working on the public documentation at the moment. PR for that will coming soon.

tomkerkhove commented on src/Promitor.Core.Scraping/Batching/AzureResourceDefinitionBatching.cs:
```diff
@@ -0,0 +1,39 @@
+        /// <see href="https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/migrate-to-batch-api?tabs=individual-response#batching-restrictions"/> 
+        /// </summary>
+        public static List<BatchScrapeDefinition<IAzureResourceDefinition>> GroupScrapeDefinitions(IEnumerable<ScrapeDefinition<IAzureResourceDefinition>> allScrapeDefinitions, int maxBatchSize) 
+        {
```
allScrapeDefinitions can be null.

hkfgo replied:
Added a check

tomkerkhove commented on src/Promitor.Core.Scraping/Batching/AzureResourceDefinitionBatching.cs:
```diff
@@ -0,0 +1,39 @@
+        /// splits the "raw" batch according to max batch size configured
+        /// </summary>
+        private static List<List<ScrapeDefinition<IAzureResourceDefinition>>> SplitScrapeDefinitionBatch(List<ScrapeDefinition<IAzureResourceDefinition>> batchToSplit, int maxBatchSize) 
+        {
```
batchToSplit can be null

tomkerkhove reviewed (approved):
LGTM, two more comments

tomkerkhove commented:
Great! We're ready to go, let me know when the CI is passing 👌

hkfgo commented:
> Great! We're ready to go, let me know when the CI is passing 👌

I think this is okay to merge as is, since we are pretty sure now that the reason for failure is test flakiness. The only failing part of CI are the Scraper OTEL Collector tests on Linux(WIndows image built just fine). I don't mind retrying every now and then but might be better to fix the flakiness through a separate PR. Wdyt?

hkfgo commented:
Oh hey the CI's passed after a few retries

hkfgo commented:
Ugh sorry I just realized I hadn't resolved all comments. Should be good to merge now

tomkerkhove commented:
Thanks a ton for all this hard work!
